### PR TITLE
Add support for more complex parallelism strategies to benchmark_litgpt.py

### PR DESCRIPTION
**Pull request by parthmannan**
## What does this PR do?

Adds support for more complex parallelism strategies - 

1. Hybrid FSDP using device mesh for PyTorch Eager/Torch.Compile
2. Bucketing for FSDP based on size for PyTorch Eager/Torch.Compile
3. Arg for bucketing for DDP based on size

cc - @carmocca @IvanYashchuk

### CODE REVIEW

**Comment by parthmannan** [comment]
@carmocca Ready for review! 🙌

File: thunder/benchmarks/benchmark_litgpt.py [hunk 1]
```diff
@@ -74,7 +77,46 @@ def __init__(
...
+        # Clarify benchmark assumptions
+        if self.sharding_size is not None:
+            assert (
+                "thunder" not in self.compile
+            ), "Hybrid Sharding (FSDP/DP) using --sharding_size is not yet supported for Thunder. Coming soon."
+
+            assert self.shard_mode in [
+                "hybrid_zero2",
+                "hybrid_zero3",
+            ], "Sharding Size is only used with Hybrid FSDP/DP style parallelism. Please "
```
Comment (carmocca) [inline thunder/benchmarks/benchmark_litgpt.py:93]: Unfinished sentence

File: thunder/benchmarks/benchmark_litgpt.py [hunk 2]
```diff
@@ -74,7 +77,46 @@ def __init__(
...
+
+            assert (
+                world_size % self.sharding_size == 0
+            ), f"World size {world_size} is not divisible by Hybrid Sharding Size {self.sharding_size}"
```
Comment (carmocca) [inline thunder/benchmarks/benchmark_litgpt.py:97]:
```suggestion
            ), f"World size {world_size} is not divisible by the sharding size {self.sharding_size}"
```

File: thunder/benchmarks/benchmark_litgpt.py [hunk 3]
```diff
@@ -74,7 +77,46 @@ def __init__(
...
+            )
+
+        assert not (
+            "thunder" in self.compile and self.bucketing_mode == "size"
+        ), "'size' bucketing mode is not supported for Thunder. Please use 'none' or 'block'."
+
+        if self.fsdp_bucket_params is not None:
+            if self.distributed_mode != "fsdp":
+                print(
+                    f"[WARNING] Found --fsdp_bucket_params but Distributed mode is {self.distributed_mode}. Will be ignnored"
```
Comment (carmocca) [inline thunder/benchmarks/benchmark_litgpt.py:112]:
```suggestion
                    f"[WARNING] Found --fsdp_bucket_params but Distributed mode is {self.distributed_mode}. Will be ignored"
```

File: thunder/benchmarks/benchmark_litgpt.py [hunk 4]
```diff
@@ -74,7 +77,46 @@ def __init__(
...
+                )
+
+            if self.bucketing_mode != "size":
+                print(
+                    f"[WARNING] Bucketing mode is set to {self.bucketing_mode}. --fsdp_bucket_params will be ignoted."
```
Comment (carmocca) [inline thunder/benchmarks/benchmark_litgpt.py:117]:
```suggestion
                    f"[WARNING] Bucketing mode is set to {self.bucketing_mode}. --fsdp_bucket_params will be ignored."
```

**Review by carmocca — APPROVED** [review]
Thank you!

File: thunder/benchmarks/benchmark_litgpt.py (hunk 4 above)
Comment (parthmannan) [reply thunder/benchmarks/benchmark_litgpt.py:117]: I can't seem to spell ignored haha! Thanks for catching these.

File: thunder/benchmarks/benchmark_litgpt.py (hunk 1 above)
Comment (parthmannan) [reply thunder/benchmarks/benchmark_litgpt.py]: Fixed.

**Pull request merged**